README.md
[![Gem Version](https://badge.fury.io/rb/openai_chatgpt.svg)](https://badge.fury.io/rb/openai_chatgpt)

# OpenaiChatgpt

OpenAI ChatGPT API is a light-weight Ruby wrapper for the Rubyists. It gives nice struct objects for completions of the chatgpt, even the raw responses are returned in nice open struct objects. It uses Faraday for HTTP requests.

> It supports all models of gpt-4 and gpt-3.5

## Installation

Install the gem and add to the application's Gemfile by executing:

    $ bundle add openai_chatgpt

If bundler is not being used to manage dependencies, install the gem by executing:

    $ gem install openai_chatgpt

---

## Usage

### API Documentation

The API documentation can be found at [https://platform.openai.com/docs/api-reference/chat/create](https://platform.openai.com/docs/api-reference/chat/create).

### Get an API key
You can get API for the chatgpt by doing the following [here](https://platform.openai.com/account/api-keys).

### Configuration

```ruby
require 'openai_chatgpt'
client = OpenaiChatgpt::Client.new(api_key: ENV['CHATGPT_API_KEY'])
```
Params:
- `api_key`: String - Required - The API key for the openai chatgpt api.

Response:
```ruby
#<OpenaiChatgpt::Client:0x00000001055fc318 @adapter=:net_http, @api_key="YOUR_API_KEY", @stubs=nil>
```

### Get completion text

```ruby
resp = OpenaiChatgpt::Client.new(api_key: "fake").completions(
  messages: [
    {role: "user", content: "Hello, how are you?"},
    {role: "assistant", content: "I am good, how are you?"},
    {role: "user", content: "What is your name?"},
  ]
)
```
Params:
- messages: Array[String] - Required - Messages to generate response for
   `[{role: "user", text: "Hello"}, {role: "bot", text: "Hi"}]`
- model: String - Optional - Default: "gpt-4" - Model to use for generating response
   ["gpt-4", "gpt-4-0314", "gpt-4-32k", "gpt-4-32k-0314", "gpt-3.5-turbo", "gpt-3.5-turbo-0301"]
- temperature: Float - Optional - Default: 1.0 - Temperature for response generation between 0.0 and 2.0
- top_p: Float - Optional - Default: 1.0 - Top p for response generation between 0.0 and 1.0
- n: Integer - Optional - Default: 1 - Number of responses to generate
- stream: Boolean - Optional - Default: false - Stream response
- stop: Array[String] - Optional - Default: nil - Stop sequence for response generation
- max_tokens: Integer - Optional - Default: 4096 - Maximum number of tokens to generate
- presence_penalty: Float - Optional - Default: 0.0 - Presence penalty for response generation between -2.0 and 2.0
- frequency_penalty: Float - Optional - Default: 0.0 - Frequency penalty for response generation between
   -2.0 and 2.0
- logit_bias: Hash - Optional - Default: nil - Logit bias for response generation
- user: String - Optional - Default: nil - User for response generation

Response:
```ruby
resp.results
=> [#<OpenaiChatgpt::Choice:0x0000000107abca68 @content="I am an AI language model created by Open", @index=0, @role="assistant">]
```
```ruby
resp.usage
=> #<OpenaiChatgpt::Usage:0x0000000107aba808 @completion_tokens=10, @prompt_tokens=36, @total_tokens=46>
```

> Also `raw_response` is available if you want the raw response from the API. This is useful if you want to do something with the raw response. Also that response is in OpenStruct Object which is a nice wrapper for the raw response and you can use 'dot' notation.

### Error Handling

```ruby
begin
  client.completions(messages: [{role: "user", text: "Hello, how are you?"}]) # this should be content instead of text
rescue OpenaiChatgpt::Error => e
  puts e.message
end
```

---

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/sulmanweb/openai_chatgpt. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/sulmanweb/openai_chatgpt/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the OpenaiChatgpt project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/sulmanweb/openai_chatgpt/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [0.2.0] - 2022-04-07

- added chatgpt 4 models "gpt-4", "gpt-4-0314", "gpt-4-32k", "gpt-4-32k-0314" by [@sulmanweb](https://github.com/sulmanweb)
- added `gpt-4` as default model by [@sulmanweb](https://github.com/sulmanweb)
- added `timeout_request` option in initializer by [@mennovergeer](https://github.com/mennovergeer)

## [0.1.2] - 2022-04-07

- increased `timeout` option to faraday connection by [@mennovergeer](https://github.com/mennovergeer)

## [0.1.1] - 2022-03-24

- added `timeout` option to faraday connection by [@mennovergeer](https://github.com/mennovergeer)

## [0.1.0] - 2022-03-06

- Initial release
- added `completions` method


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "openai_chatgpt"
require "minitest/autorun"
require "faraday"
require "json"

# :nodoc:
class Minitest::Test # rubocop:disable Style/ClassAndModuleChildren
  def stub_response(fixture:, status: 200, headers: { "Content-Type" => "application/json" })
    [status, headers, File.read("test/fixtures/#{fixture}.json")]
  end

  def stub_request(path, response:, method: :get, body: {})
    Faraday::Adapter::Test::Stubs.new do |stub|
      arguments = [method, "/#{path}"]
      arguments << body.to_json if %i[post put patch].include?(method)
      stub.send(*arguments) { |_env| response }
    end
  end
end


test/openai_chatgpt/client_test.rb
# frozen_string_literal: true

require "test_helper"

# :nodoc:
class ClientTest < Minitest::Test
  def test_api_key
    client = OpenaiChatgpt::Client.new api_key: "test"
    assert_equal "test", client.api_key
  end
end


lib/openai_chatgpt/client.rb
# frozen_string_literal: true

module OpenaiChatgpt
  # Client having API methods and api_key
  class Client
    # Base url for OpenAI ChatGPT API
    BASE_URL = "https://api.openai.com/v1/chat/completions"
    attr_reader :api_key, :adapter

    # #### Params:
    # api_key: String - Required - API key for OpenAI ChatGPT API (https://platform.openai.com/account/api-keys)
    # adapter: Symbol - Optional - Default: :net_http - Adapter for Faraday connection
    # stubs: Faraday::Adapter::Test::Stubs - Optional - Default: nil - Stubs for Faraday connection
    # request_timeout: Integer - Optional - Default: 1200 - Timeout for Faraday connection
    # #### Example:
    # client = OpenaiChatgpt::Client.new api_key: "test"
    # #### Description:
    # Client for OpenAI ChatGPT API
    def initialize(api_key:, adapter: Faraday.default_adapter, stubs: nil, request_timeout: 1200)
      @api_key = api_key
      @adapter = adapter
      @request_timeout = request_timeout

      # Test stubs for requests
      @stubs = stubs
    end

    # #### Params:
    # messages: Array[String] - Required - Messages to generate response for
    #   [{role: "user", text: "Hello"}, {role: "bot", text: "Hi"}]
    # model: String - Optional - Default: "gpt-4" - Model to use for generating response
    #   ["gpt-4", "gpt-4-0314", "gpt-4-32k", "gpt-4-32k-0314", "gpt-3.5-turbo", "gpt-3.5-turbo-0301"]
    # temperature: Float - Optional - Default: 1.0 - Temperature for response generation between 0.0 and 2.0
    # top_p: Float - Optional - Default: 1.0 - Top p for response generation between 0.0 and 1.0
    # n: Integer - Optional - Default: 1 - Number of responses to generate
    # stream: Boolean - Optional - Default: false - Stream response
    # stop: Array[String] - Optional - Default: nil - Stop sequence for response generation
    # max_tokens: Integer - Optional - Default: 4096 - Maximum number of tokens to generate
    # presence_penalty: Float - Optional - Default: 0.0 - Presence penalty for response generation between -2.0 and 2.0
    # frequency_penalty: Float - Optional - Default: 0.0 - Frequency penalty for response generation between
    #   -2.0 and 2.0
    # logit_bias: Hash - Optional - Default: nil - Logit bias for response generation
    # user: String - Optional - Default: nil - User for response generation
    # #### Example:
    # OpenaiChatgpt::Client.new(api_key: "fake")
    #   .completions(messages: [{role: "user", content: "Hello"}, {role: "assistant", content: "Hi"}])

    # #### Description:
    # completions method for OpenAI ChatGPT API
    # #### Response:
    # Response object with choices
    # #### Error:
    # OpenaiChatgpt::Error
    def completions( # rubocop:disable Metrics/MethodLength, Metrics/ParameterLists
      messages:,
      model: "gpt-4",
      temperature: 1.0,
      top_p: 1.0,
      n: 1, # rubocop:disable Naming/MethodParameterName
      stream: false,
      stop: nil,
      max_tokens: 2000,
      presence_penalty: 0,
      frequency_penalty: 0,
      logit_bias: {},
      user: ""
    )
      Response.new post("", {
        model: model,
        messages: messages,
        temperature: temperature,
        top_p: top_p,
        n: n,
        stream: stream,
        stop: stop,
        max_tokens: max_tokens,
        presence_penalty: presence_penalty,
        frequency_penalty: frequency_penalty,
        logit_bias: logit_bias,
        user: user
      }.to_json)
    end

    private

    # makes a get request to the given path with params
    def post(path, params = {})
      handle_response connection.post(path, params)
    end

    # handles the response from the api and raises error if any
    # returns the response if no error
    # Error codes: https://platform.openai.com/docs/guides/error-codes/api-errors
    def handle_response(response)
      return response if response.status == 200

      raise Error, "#{response.status}: #{response.body["error"]["message"]}"
    end

    # creates a faraday connection with base url and api_key
    def connection
      @connection ||= Faraday.new(BASE_URL) do |conn|
        conn.headers["Authorization"] = "Bearer #{@api_key}"
        conn.headers["Content-Type"] = "application/json"
        conn.options.timeout = @request_timeout
        conn.response :json, content_type: "application/json"
        conn.adapter adapter, @stubs
      end
    end
  end
end
